api/model/src/main/java/org/projectnessie/model/Util.java

```diff
@@ -44,15 +53,91 @@ private Util() {}
   public static final char REF_HASH_SEPARATOR = '@';
 
   /**
-   * Convert from path encoded string to normal string.
+   * Convert from path encoded string to normal string, supports all Nessie Spec versions.
+   *
+   * <p>The {@code encoded} parameter is split at dot ({@code .}) characters.
+   *
+   * <p>Legacy compatibility: a dot ({@code .}) character can be represented using ASCII 31 (0x1F)
+   * or ASCII 0 (0x00).
+   *
+   * <p>Escaping, for compatibility w/ <a
+   * href="https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0.html#uri-path-canonicalization">Jakarta
+   * Servlet Specification 6, URI Path Canonicalization</a> and <a
+   * href="https://datatracker.ietf.org/doc/html/rfc3986#section-3.3">RFC 3986, section 3.3</a>.
+   * This is available with Nessie Spec version 2.2.0, as advertised via {@link
+   * NessieConfiguration#getSpecVersion()}.
+   *
+   * <ul>
+   *   <li>Fact: two consecutive dots ({@code ..}) would represent an <em>empty</em> namespace
+   *       elements, which is illegal. It is correct to assume that this character sequence does not
+   *       occur in encoded content keys.
+   *   <li>The sequence {@code ..*} is the encoded representation for a single dot character {@code
+   *       .}.
+   *   <li>The sequence {@code ..^} is the encoded representation for a slash {@code /}, which is a
+   *       URI path separator.
+   *   <li>The sequence {@code ..=} is the encoded representation for a backslash {@code \}, which
+   *       is a URI path separator.
+   *   <li>The sequence {@code ..#} is the encoded representation for a hash {@code #}, which is a
+   *       URI path separator.
+   *   <li>The sequence {@code ..} followed by any character not mentioned in the sequences above,
+   *       means that the first dot represents an element boundary, decoding should continue after
+   *       the first {@code .} character.
+   * </ul>
    *
    * @param encoded Path encoded string
    * @return Actual key.
    */
   public static List<String> fromPathString(String encoded) {
-    return Arrays.stream(encoded.split("\\."))
-        .map(x -> x.replace(GROUP_SEPARATOR, DOT).replace(ZERO_BYTE, DOT))
-        .collect(Collectors.toList());
+    List<String> elements = new ArrayList<>();
+    int l = encoded.length();
+    StringBuilder e = new StringBuilder();
+    for (int i = 0; i < l; i++) {
+      char c = encoded.charAt(i);
+      switch (c) {
+        case DOT:
+          if (encoded.charAt(i + 1) == DOT) {
```

Comment:
Could this hit a `StringIndexOutOfBoundsException`?

Reply:
s..t in, s..t out ;)